drivers/source/SerialBase.cpp

```diff
@@ -164,6 +164,19 @@ void SerialBase::_deinit()
     serial_free(&_serial);
 }
 
+int SerialBase:: sync()
+{
+    lock();
+
+    while (!serial_tx_empty(&_serial)) {
+        // See send_break()
+        wait_us(18000000 / _baud);
```

Comment:
At the beginning i was thinking to add a counter to avoid infinite loops if something goes wrong in the serial peripheral. So i wanted to use the delay to keep track of the byte time. 

Unfortunately the counter solution is not working if flow control is used so i've not implemented it and did not thought more about the delay.

At this point, considering the flow control, i think the best solution is to use ThisThread::sleep(1ms) as you suggested.

Reply:
Hmm, I guess if you wanted to implement a timeout, that would be a reason to keep the error return in `SerialBase`.

And I guess a timeout is reasonable.

I wonder how flow control should really be handled. There is an argument for sync returning when the serial TX is quiescent, rather than drained, so in a flow control setup the other end can't stop you sleeping.

Just having a timeout would approximate the behaviour. You'd only hit the timeout if flow control was holding you up, so you'd likely be quiescent at timeout.

Only problem is either way, it's racy. The other end could choose to release flow control just as sync returns, making you active again. Having it simply be timed with a "timeout" error indicates the danger, distinguising the "drained" (safe) and "holding" states.

Reply:
Yes, timeout is reasonable but not a robust solution and it may over complicate the sync behavior. Also the implementation it is not so fast because it depends on the hardware buffer size of each target.

I think we have a list of possible solutions with flow control:
 1 do nothing end exit with an error
 2 wait and check if receiver is holding the transmission and, in that case, exit with an error
 3 block the caller and wait until all byte are sent even if we have to wait forever or for a long time (actual solution)

1 and 3 are easy I will verify if there is an easy way to check if the receiver is holding the transmission so we could exit from sync with an appropriate error.

Reply:
Most of these embedded devices have a 1 byte buffer. Never seen anything bigger than a 16-byte FIFO in a non-embedded system. So without flow control, you're talking ~2 seconds tops at 75 baud, and less than 20ms at 9600.

I see no need to try to fine-tune this. Even one second timeout covers non-flow-blocked drain in any realistic case.

Reply:
oh that is great, in this case it is really worth to implement a timeout directly into SerialBase::sync based on the 16-byte FIFO limit. I don't know why but i was convinced the Renesas target had much more bigger FIFO, but i've double checked and i was wrong.

So i will verify point 2 and update the pr

Reply:
I did not found any proper way to easily check if the receiver is holding the transmission and my last thought on it is that if flow control is used sync function should rely on flow control impelmentation and block until transmission is done without timeout. See c25ae4934a52b2e8a0b964ba47424bca0bbc7a01